boot/cypress/MCUBootApp/encrypted_mbedtls.c

```diff
@@ -0,0 +1,457 @@
+    }
+
+    memset( counter, 0, MBEDTLS_AES_BLOCK_SIZE);
+
+    memset( stream_block, 0, MBEDTLS_AES_KEY_SIZE );
+    rc = mbedtls_aes_crypt_ctr( &aes_ctx, MBEDTLS_AES_KEY_SIZE, &nc_off, counter, stream_block,
+                           &buf[EC_CIPHERKEY_INDEX], enckey );
+    if (rc != 0) {
+        mbedtls_aes_free( &aes_ctx );
+        return -1;
+    }
+
+    mbedtls_aes_free( &aes_ctx );
+    
+    rc = 0;
+
+    return rc;
+}
+
+/*
+ * Load encryption key.
+ */
+int
+boot_enc_load(struct enc_key_data *enc_state, int image_index,
+        const struct image_header *hdr, const struct flash_area *fap,
+        struct boot_status *bs)
+{
+    uint32_t off;
+    uint16_t len;
+    struct image_tlv_iter it;
+    uint8_t buf[EXPECTED_ENC_LEN];
+    uint8_t slot;
+    int rc;
+    
+    rc = flash_area_id_to_multi_image_slot(image_index, fap->fa_id);
+    if (rc < 0) {
+        return rc;
+    }
+    slot = rc;
+
+    /* Already loaded... */
+    if (enc_state[slot].valid) {
+        return 1;
+    }
+
+    rc = bootutil_tlv_iter_begin(&it, hdr, fap, EXPECTED_ENC_TLV, false);
+    if (rc != 0) {
+        return -1;
+    }
+
+    rc = bootutil_tlv_iter_next(&it, &off, &len, NULL);
+    if (rc != 0) {
+        return rc;
+    }
+
+    if (len != EXPECTED_ENC_LEN) {
+        return -1;
+    }
+
+    rc = flash_area_read(fap, off, buf, EXPECTED_ENC_LEN);
+    if (rc != 0) {
+        return -1;
+    }
+
+    return boot_enc_decrypt(buf, bs->enckey[slot]);
+}
+
+bool
+boot_enc_valid(struct enc_key_data *enc_state, int image_index,
+        const struct flash_area *fap)
+{
+    int rc;
+
+    rc = flash_area_id_to_multi_image_slot( image_index, fap->fa_id );
+    if (rc < 0) {
+        /* can't get proper slot number - skip encryption, */
+        /* postpone the error for a upper layer */
+        return false;
+    }
+
+    return enc_state[rc].valid;
+}
+
+void
+boot_encrypt(struct enc_key_data *enc_state, int image_index,
+        const struct flash_area *fap, uint32_t off, uint32_t sz,
+        uint32_t blk_off, uint8_t *buf)
+{
+    struct enc_key_data *enc;
+    uint32_t i, j;
+    uint8_t u8;
+    uint8_t nonce[16];
+    uint8_t blk[16];
+    int rc;
+
+    memset( nonce, 0, 12);
+    off >>= 4;
+    nonce[12] = (uint8_t)(off >> 24);
+    nonce[13] = (uint8_t)(off >> 16);
+    nonce[14] = (uint8_t)(off >> 8);
+    nonce[15] = (uint8_t)off;
+
+    rc = flash_area_id_to_multi_image_slot( image_index, fap->fa_id );
+    if (rc < 0) {
+        assert(0);
+        return;
+    }
+
+    enc = &enc_state[rc];
+    assert( enc->valid == 1);
+    for (i = 0; i < sz; i++) {
+        if (i == 0 || blk_off == 0) {
+            mbedtls_aes_crypt_ecb( &enc->aes_ctr, MBEDTLS_AES_ENCRYPT, nonce, blk );
+
+            for (j = 16; j > 0; --j) {
+                if (++nonce[j - 1] != 0) {
+                    break;
+                }
+            }
+        }
+
+        u8 = *buf;
+        *buf++ = u8 ^ blk[blk_off];
+        blk_off = (blk_off + 1) & 0x0f;
```

Comment:
This block is not required because MbedTLS has proper streaming for CTR with `mbedtls_aes_crypt_ctr`